File: js(ts)-solutions/1415TheKthLexicographicalStringOfAllHappyStringsOfLengthN.ts
function getHappyString(n: number, k: number): string {
    let happyStrings: string[] = [];
    const dfs = (stack: string[]) => {
        if (stack.length === n) {
            happyStrings.push(stack.join(""));
            return
        }

        for (const letter of "abc") {
            if (letter !== stack.at(-1)) dfs([...stack, letter]);
        }
    };

    for (const letter of "abc") dfs([letter]);

    happyStrings.sort();

    return (k <= happyStrings.length) ? happyStrings[k - 1] : "";
};

File: leetcode-questions/1415-The-k-th-lexicographical-string-of-All-Happy-Strings-of-Length-n/1415-The-k-th-lexicographical-string-of-All-Happy-Strings-of-Length-n.py
class Solution:
    def __init__(self):
        self.happy_strings = []
        self.N = 0

    def dfs(self, stack):
        if len(stack) == self.N:
            self.happy_strings.append("".join(stack))
            return

        for letter in "abc":
            if letter != stack[-1]:
                stack.append(letter)
                self.dfs(stack)
                stack.pop()

    def getHappyString(self, n: int, k: int) -> str:
        self.N = n
        for letter in "abc":
            self.dfs([letter])

        self.happy_strings.sort()
        return "" if k > len(self.happy_strings) else self.happy_strings[k - 1]
